Initial implementation of WebGLSync

imiklos opened the pull request:
This patch adds initial support for [WebGLSync](https://www.khronos.org/registry/webgl/specs/latest/2.0/#3.7.14).

Note:
There is no test for the isSync, deleteSync and waitSync functions in the `conformance2/sync/sync-webgl-specific.html`.

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors

<!-- Either: -->
- [X] There are tests for these changes

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/24250)
<!-- Reviewable:end -->

Commit:
Initial implementation of WebGLSync

This patch adds initial support for WebGLSync.

Note:
There is no test for the isSync, deleteSync and waitSync functions in the `conformance2/sync/sync-webgl-specific.html`.

highfive commented:
Thanks for the pull request, and welcome! The Servo team is excited to review your changes, and you should hear from @manishearth (or someone else) soon.

highfive commented:
Heads up! This PR modifies the following files:
 * @asajeffrey: components/script/dom/webglrenderingcontext.rs, components/script/dom/webidls/WebGL2RenderingContext.webidl, components/script/dom/webidls/WebGLSync.webidl, components/script/dom/mod.rs, components/script/dom/bindings/trace.rs and 2 more
 * @jgraham: tests/wpt/webgl/meta/conformance2/sync/sync-webgl-specific.html.ini
 * @kichjang: components/script/dom/webglrenderingcontext.rs, components/script/dom/webidls/WebGL2RenderingContext.webidl, components/script/dom/webidls/WebGLSync.webidl, components/script/dom/mod.rs, components/script/dom/bindings/trace.rs and 2 more

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!

jdm commented on components/script/dom/webgl2renderingcontext.rs:
```diff
@@ -1044,6 +1052,123 @@ impl WebGL2RenderingContextMethods for WebGL2RenderingContext {
+    /// https://www.khronos.org/registry/webgl/specs/latest/2.0/#3.7.14
+    fn ClientWaitSync(&self, sync: &WebGLSync, flags: u32, timeout: u64) -> u32 {
+        if !sync.is_valid() {
+            return constants::WAIT_FAILED;
```
We need to set an InvalidOperation webgl error here.

jdm commented on components/script/dom/webgl2renderingcontext.rs:
```diff
@@ -1044,6 +1052,123 @@ impl WebGL2RenderingContextMethods for WebGL2RenderingContext {
+            constants::SYNC_STATUS => {
+                let (sender, receiver) = webgl_channel().unwrap();
+                self.base
+                    .send_command(WebGLCommand::GetSyncParameter(sync.id(), pname, sender));
```
We will need to do something more clever here - "Repeatedly fetching a sync object's SYNC_STATUS parameter in a loop, without returning control to the user agent, must always return the same value." We should be able to use a similar strategy to what I described in https://github.com/servo/servo/pull/24178.

imiklos replied:
You are right! I will update the PR with this change.
I've just noticed that the test file (`sync-webgl-specific.html`) has a wrong check on the `getSyncParameter` return value which has to be `gl.SIGNALED` instead of `gl.SYNC_SIGNALED`, otherwise the test will always pass.

jdm commented on components/script/dom/webglsync.rs:
```diff
@@ -0,0 +1,93 @@
+
+impl WebGLSync {
+    pub fn client_wait_sync(&self, context: &WebGLRenderingContext, flags: u32, timeout: u64) {
+        if self.get_client_sync_status().is_none() {
```
If the timeout is hit and calling code wants to wait again, won't this return incorrect results the second time?

imiklos replied:
I think the second result will be incorrect, I should check if the status is `TIMEOUT_EXPIRED` or `WAIT_FAILED`.

jdm commented on components/script/dom/webglsync.rs:
```diff
@@ -0,0 +1,93 @@
+    pub fn id(&self) -> WebGLSyncId {
+        self.sync_id
+    }
+}
```
I think we need to add a Drop implementation that calls delete: "Note that underlying GL object will be automatically marked for deletion when the JS object is destroyed"

jdm commented:
Also some build errors:
```
[0m[1m[38;5;9merror[E0061][0m[0m[1m: this function takes 5 parameters but 2 parameters were supplied[0m
[0m    [0m[0m[1m[38;5;12m--> [0m[0mcomponents/canvas/webgl_thread.rs:1383:38[0m
[0m     [0m[0m[1m[38;5;12m|[0m
[0m[1m[38;5;12m1383[0m[0m [0m[0m[1m[38;5;12m| [0m[0m                let value = ctx.gl().get_sync_iv(sync_id.get() as *const _, param);[0m
[0m     [0m[0m[1m[38;5;12m| [0m[0m                                     [0m[0m[1m[38;5;9m^^^^^^^^^^^[0m[0m [0m[0m[1m[38;5;9mexpected 5 parameters[0m

[0m[1m[38;5;9merror[E0608][0m[0m[1m: cannot index into a value of type `()`[0m
[0m    [0m[0m[1m[38;5;12m--> [0m[0mcomponents/canvas/webgl_thread.rs:1384:29[0m
[0m     [0m[0m[1m[38;5;12m|[0m
[0m[1m[38;5;12m1384[0m[0m [0m[0m[1m[38;5;12m| [0m[0m                sender.send(value[0] as u32).unwrap();[0m
[0m     [0m[0m[1m[38;5;12m| [0m[0m                            [0m[0m[1m[38;5;9m^^^^^^^^[0m[0m [0m[0m[1m[38;5;9mhelp: to access tuple elements, use: `value.0`[0m

[0m[1m[38;5;9merror[0m[0m[1m: aborting due to 2 previous errors[0m
```

jdm commented:
This will need to use `cargo update -p sparkle`.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #24242) made this pull request unmergeable. Please resolve the merge conflicts.

imiklos commented:
Thank you for your review! I've updated the pull request. Added drop implementation and task queue for the sync_status fetching as you mentioned before.

jdm commented:
@bors-servo r+
Thanks!

bors-servo commented:
:pushpin: Commit fb043a3 has been approved by `jdm`

<!-- @bors-servo r=jdm fb043a3bf7f675c1038417200fa7e5407f69f109 -->

bors-servo commented:
:hourglass: Testing commit fb043a3bf7f675c1038417200fa7e5407f69f109 with merge 68d3cbc44a1e08d95a1c7fb6afa1785f42518ee5...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/13541)

jdm commented on components/script/dom/webidls/WebGLSync.webidl:
```diff
@@ -0,0 +1,10 @@
+// https://www.khronos.org/registry/webgl/specs/latest/2.0/#3.7.14
+//
+
+interface WebGLSync : WebGLObject {
```
This needs to be gated on the WebGL2 preference: https://github.com/servo/servo/blob/f63b404e0cbf30380c4043700861110d06e548bb/components/script/dom/webidls/WebGL2RenderingContext.webidl#L58

jdm reviewed (changes requested):
The test results for the idlharness test in web-platform-tests/webgl/ will also need to be updated.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #24178) made this pull request unmergeable. Please resolve the merge conflicts.

imiklos commented:
Thanks, I've updated the pull request with the `methods-2` and `Pref` changes, but I couldn't reproduce the `idlharness` fails.

jdm commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 248545d has been approved by `jdm`

<!-- @bors-servo r=jdm 248545ddda503e06bc59b5274c63a6c25da4b355 -->

bors-servo commented:
:hourglass: Testing commit 248545ddda503e06bc59b5274c63a6c25da4b355 with merge 13a43e69e0fa16540ec02b6fc3569202470e9e5d...

bors-servo commented:
:sunny: Test successful - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/13572), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/13634), [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/bk6U3dmLQXOsjM2jn-VtSQ)
Approved by: jdm
Pushing 13a43e69e0fa16540ec02b6fc3569202470e9e5d to master...
